### PR TITLE
Cleanup PHPSTAN config file + needed fixes

### PR DESCRIPTION
**Pull request by mollux**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:

This PR improves the PHPSTAN config and removes as much as possible the excluded files and ignored errors + add the needed code fixes.

the PHPSTAN check failure is normal, as there is 1 changed line (lowered the occurrences from 2 -> 1)

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2.

### CODE REVIEW

File: app/bundles/CoreBundle/Entity/UpsertTrait.php [hunk 1]
```diff
@@ -1,37 +0,0 @@
-<?php
-
-declare(strict_types=1);
-
-namespace Mautic\CoreBundle\Entity;
-
-trait UpsertTrait
```
Comment (escopecz) [inline app/bundles/CoreBundle/Entity/UpsertTrait.php]: Removing this will break the Custom Objects plugin. See https://github.com/mautic/mautic/pull/11746

Comment (mollux) [reply app/bundles/CoreBundle/Entity/UpsertTrait.php]: so, there is a file in Mautic only used for potential plugins? interesting :) 

Wouldn't it be more logical to make a `app/bundles/CoreBundle/Entity/UpsertBase.php` class that implements the interface? and that base class is extended in plugins?
Or is here is a reason this is a trait?

Comment (mollux) [reply app/bundles/CoreBundle/Entity/UpsertTrait.php]: as this discussion is OOS for this PR, I'll address the issue to unblock it

**Comment by mollux** [comment]
@escopecz I addressed the `UpsertTrait` issue, see https://github.com/mautic/mautic/pull/12225/commits/f97590f94d05c9128dbf1e6db52db9214a9c67b4#diff-995edee38ad4f8387e58ebd52c31bcc04c56cc2448d331b1cf5e0b35c57b9efa

**Review by mabumusa1 — DISMISSED** [review]
Changes looks good, thanks @mollux

File: phpstan-bootstrap.php [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<?php
+
+define('MAUTIC_ENV', '');
```
Comment (mabumusa1) [inline phpstan-bootstrap.php:3]: @lenonleite there will be an upcoming change to `phpstan`, this change will be impacted with your work, so I thought of letting you know

Comment (mollux) [reply phpstan-bootstrap.php:3]: @mabumusa1 did you tag the wrong person?

Comment (mabumusa1) [reply phpstan-bootstrap.php:3]: No, @lenonleite is working PR #12206, there is a lot going on with that PR regarding `.env` so I thought of giving him heads up about this change

**Review by escopecz — DISMISSED** [review]
Great improvement. Thanks!

File: app/bundles/CoreBundle/Tests/_support/FunctionalTester.php [hunk 3]
```diff
@@ -1,28 +0,0 @@
...
- * @method void                    expect($prediction)
- * @method void                    amGoingTo($argumentation)
- * @method void                    am($role)
- * @method void                    lookForwardTo($achieveValue)
- * @method void                    comment($description)
- * @method \Codeception\Lib\Friend haveFriend($name, $actorClass = NULL)
- *
- * @SuppressWarnings(PHPMD)
- */
-class FunctionalTester extends \Codeception\Actor
```
Comment (escopecz) [inline app/bundles/CoreBundle/Tests/_support/FunctionalTester.php:21]: I was looking at where this file come from. It looks like it's a start to support Codeception tests. See the readme changes in https://github.com/mautic/mautic/pull/4868

It looks like the codeception library was since removed from composer.json so it's not useable. I'm wondering whether we shouldn't remove the rest. Like the functional.suite.yml file.

File: app/bundles/CoreBundle/Entity/UpsertTrait.php (hunk 1 above)
Comment (escopecz) [reply app/bundles/CoreBundle/Entity/UpsertTrait.php]: > so, there is a file in Mautic only used for potential plugins? interesting :)

It's more complicated than that. It exists in Acquia's fork of Mautic and serves another purpose in the core. However, this change was not pushed to Mautic OSS yet.

The Custom Objects plugin was released as OSS last year. The community started using it. Then we pushed some update that broke it for the community and so we quickly patched it with adding this trait to the next Mautic version so we'd fix it. I have GH Actions CI in progress for the plugin to ensure it won't happen, but there are other priorities.

So this was a quick fix and the intentional PR where it's used in the core is coming soon. It will all make sense in the end.

TBH I'm not clear on whether inheritance is less evil than traits. Both have pros and cons.

File: app/bundles/CoreBundle/Tests/_support/FunctionalTester.php (hunk 3 above)
Comment (mollux) [reply app/bundles/CoreBundle/Tests/_support/FunctionalTester.php:21]: it's not really related to phpstan, but it's cleanup :) So I added that in https://github.com/mautic/mautic/pull/12225/commits/27bca7df749bf87f4bb78c57c2220793c8d37ab7

**Comment by dadarya0** [comment]
@mollux please fix Mautic tests and validations / PHPStan baseline changes - 8.0

File: tests/Mautic/StageBundle/Tests/Entity/StageRepositoryTest.php [hunk 4]
```diff
@@ -1,137 +0,0 @@
...
-namespace Mautic\StageBundle\Tests\Entity;
-
-use Doctrine\ORM\AbstractQuery;
-use Doctrine\ORM\EntityManager;
-use Doctrine\ORM\Mapping\ClassMetadata;
-use Doctrine\ORM\QueryBuilder;
-use Mautic\StageBundle\Entity\Stage;
-use Mautic\StageBundle\Entity\StageRepository;
-
-class StageRepositoryTest extends \PHPUnit\Framework\TestCase
```
Comment (escopecz) [inline tests/Mautic/StageBundle/Tests/Entity/StageRepositoryTest.php:14]: This one doesn't seem to be related to Codeception. I realize now that tests like this one are not helpful and a short functional test would do a better job. But do we want to throw it away?

**Comment by mollux** [comment]
@dadarya0 see the intro 

```
the PHPSTAN check failure is normal, as there is 1 changed line (lowered the occurrences from 2 -> 1)
```

File: tests/Mautic/StageBundle/Tests/Entity/StageRepositoryTest.php (hunk 4 above)
Comment (mollux) [reply tests/Mautic/StageBundle/Tests/Entity/StageRepositoryTest.php:14]: that test hasn't been executed for a long time, as it's not on a folder that phpunit checks.

see the number of tests before and after that commit

[before](https://github.com/mautic/mautic/actions/runs/4743304878/jobs/8422682201):

```
OK, but incomplete, skipped, or risky tests!
Tests: 3325, Assertions: 9967, Skipped: 5.
```

[after](https://github.com/mautic/mautic/actions/runs/4746183342/jobs/8429482783)
```
OK, but incomplete, skipped, or risky tests!
Tests: 3325, Assertions: 9967, Skipped: 5.
```

Comment (mollux) [reply tests/Mautic/StageBundle/Tests/Entity/StageRepositoryTest.php:14]: So re-adding it would be another PR, tis is the cleanup

Comment (escopecz) [reply tests/Mautic/StageBundle/Tests/Entity/StageRepositoryTest.php:14]: Ah, I see. It's in a wrong folder. I'm OK with removing it then. It tests implementation and not I/O so it's close to useless anyway.

**Review by escopecz — APPROVED** [review]
Thank you! 👍
